File: mobile_app/lib/models/ai_agent.dart
enum AgentActionStatus {
  NONE,
  PENDING,
  APPROVED,
  REJECTED,
  EXECUTED
}

class AgentMessage {
  final String id;
  final String sessionId;
  final String role;
  final String? content;
  final List<dynamic>? toolCalls;
  final String? toolCallId;
  final AgentActionStatus actionStatus;
  final Map<String, dynamic>? actionData;
  final DateTime createdAt;

  AgentMessage({
    required this.id,
    required this.sessionId,
    required this.role,
    this.content,
    this.toolCalls,
    this.toolCallId,
    required this.actionStatus,
    this.actionData,
    required this.createdAt,
  });

  factory AgentMessage.fromJson(Map<String, dynamic> json) {
    return AgentMessage(
      id: json['id'],
      sessionId: json['session_id'],
      role: json['role'],
      content: json['content'],
      toolCalls: json['tool_calls'],
      toolCallId: json['tool_call_id'],
      actionStatus: _parseStatus(json['action_status']),
      actionData: json['action_data'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  static AgentActionStatus _parseStatus(String? status) {
    switch (status) {
      case 'PENDING': return AgentActionStatus.PENDING;
      case 'APPROVED': return AgentActionStatus.APPROVED;
      case 'REJECTED': return AgentActionStatus.REJECTED;
      case 'EXECUTED': return AgentActionStatus.EXECUTED;
      default: return AgentActionStatus.NONE;
    }
  }
}

class AgentSession {
  final String id;
  final String trainerId;
  final String? title;
  final List<AgentMessage> messages;
  final DateTime createdAt;
  final DateTime updatedAt;

  AgentSession({
    required this.id,
    required this.trainerId,
    this.title,
    required this.messages,
    required this.createdAt,
    required this.updatedAt,
  });

  factory AgentSession.fromJson(Map<String, dynamic> json) {
    return AgentSession(
      id: json['id'],
      trainerId: json['trainer_id'],
      title: json['title'],
      messages: (json['messages'] as List? ?? [])
          .map((m) => AgentMessage.fromJson(m))
          .toList(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }
}

File: backend/app/models/ai_agent.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, ForeignKey, Enum, DateTime, JSON
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from app.db.session import Base
import enum

class AgentActionStatus(str, enum.Enum):
    NONE = "NONE"
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    EXECUTED = "EXECUTED"

class AgentSession(Base):
    __tablename__ = "ai_agent_sessions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    trainer_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    title = Column(String, nullable=True) # Ex: "Criação de treinos para João"

    trainer = relationship("User", foreign_keys=[trainer_id], backref="agent_sessions")
    student = relationship("User", foreign_keys=[student_id])
    messages = relationship("AgentMessage", back_populates="session", cascade="all, delete-orphan", order_by="AgentMessage.created_at")

from sqlalchemy import event

@event.listens_for(AgentSession, 'before_update')
def receive_before_update(mapper, connection, target):
    target.updated_at = datetime.utcnow()

class AgentMessage(Base):
    __tablename__ = "ai_agent_messages"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(UUID(as_uuid=True), ForeignKey("ai_agent_sessions.id", ondelete="CASCADE"), nullable=False)
    role = Column(String, nullable=False)  # user, assistant, tool
    content = Column(String, nullable=True) # Texto da mensagem
    
    # Para Function Calling
    tool_calls = Column(JSONB, nullable=True) # Se a IA solicitou uma função
    tool_call_id = Column(String, nullable=True) # Se esta mensagem for a resposta de uma tool
    
    # Estado da Ação (Rich UI)
    action_status = Column(Enum(AgentActionStatus), default=AgentActionStatus.NONE, nullable=False)
    action_data = Column(JSONB, nullable=True) # Payload da ação (ex: JSON do treino planejado)
    
    created_at = Column(DateTime, default=datetime.utcnow)

    session = relationship("AgentSession", back_populates="messages")
